### PR TITLE
[ramda] add more general overload signature for pickBy

### PR DESCRIPTION
**Pull request by jituanlin**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [ ] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## ramda ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ramda@latest/))
was missing the following properties:
1. sequence
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 2e7345337f853b50178dee0948d44a98dd4a28ba
</p>

File: types/ramda/index.d.ts [hunk 1]
```diff
@@ -1605,6 +1605,8 @@ export function pickAll(names: readonly string[]): <T, U>(obj: T) => U;
 /**
  * Returns a partial copy of an object containing only the keys that satisfy the supplied predicate.
  */
+export function pickBy(pred: ObjPred, record: Record<keyof any, any>): Record<keyof any, any>;
```
Comment (Nicholaiii) [inline types/ramda/index.d.ts:1608]: ObjPred takes a defaulted generic, you should use it.

File: types/ramda/index.d.ts [hunk 2]
```diff
@@ -1605,6 +1605,8 @@ export function pickAll(names: readonly string[]): <T, U>(obj: T) => U;
...
+export function pickBy(pred: ObjPred): (record: Record<keyof any, any>) => Record<keyof any, any>;
```
Comment (Nicholaiii) [inline types/ramda/index.d.ts:1609]: Why are you using any instead of unknown? Any is less safe.

**Review by Nicholaiii — CHANGES_REQUESTED** [review]
These just looks like weaker typed versions of the existing, whats the reason?

**Comment by jituanlin** [comment]
> These just looks like weaker typed versions of the existing, whats the reason?
This type of signature is deliberately designed to be "weaker"(wider):
1. According to type signature of official document `((v, k) → Boolean) → {k: v} → {k: v}`, the input and output of `pickBy` are `Record` 
2. The previous type signature forced the user to specify type parameters explicitly, which made it awkward to use `pickBy` in the `pipe`. `export function pickBy<T>(pred: ObjPred<T>): <U, V extends T>(obj: V) => U;` `U` will be unknown in curring case and the only way to narrow it is to assert the entire curried function type signature.

File: types/ramda/index.d.ts (hunk 1 above)
Comment (jituanlin) [reply types/ramda/index.d.ts:1608]: Yes, using type parameters in this case will make type stronger, but it breaks the forward compatibility:
The previous type signature forced the user to specify type parameters implicitly. In order to make type overload work better and not break the already existing use case, I design the new signature to not use type parameter.

File: types/ramda/index.d.ts (hunk 2 above)
Comment (jituanlin) [reply types/ramda/index.d.ts:1609]: Yes, `unknown` is better, I will fix that.

Comment (jituanlin) [reply types/ramda/index.d.ts:1609]: No, I am sure that `any` is better(If you try to it with `filter` in `pipe`).
```ts
export interface Dictionary<A> {
    [index: string]: A;
}
```

Do you feel that this project tried to describe the type accurately so hard, but in the end it made it terrible?

Anytime, at any time, we have written a *correct* code, but what we get is a type error, and then we explicitly assert the type.

Why do we do this on earth only to make our job harder?

This project is so crazy that using Typescript  to describe Clojure like Haskell.

No! We don't need the *perfect* type, just make things work!

**Comment by elibarzilay** [comment]
Would be good to get a review from an owner.

File: types/ramda/index.d.ts (hunk 1 above)
Comment (gabritto) [reply types/ramda/index.d.ts:1608]: Could you clarify what is the issue with the previous signatures and how the new ones are an improvement? I am still confused after reading the code and the discussions.

**Comment by amcasey** [comment]
I'm reading the change as a desire to be able to call `pickBy` without a type argument and get a looser type back.  That seems reasonable, but I haven't seen any support for the idea, much less for this particular way of doing it.  For a package this popular, that's a dealbreaker.  Sorry.

**Pull request closed without merging**